### PR TITLE
[Test] Full version of use case testing, all use cases have passed the tests

### LINKED ISSUE #31: [Test] Usecase Test

### PR DESCRIPTION
**Pull request by josephpcsywm**
Demonstrated comprehensive testing of program functionality, specifically use cases. Tests validated accurate data storage and covered various scenarios, including corner cases, embodying the team's thorough approach to verification.

### CODE REVIEW

File: src/test/java/usecasetesting/AskQuestionUseCaseTest.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    final static int ATTORNEY_ID = 11345678;
+    final static int SECOND_ATTORNEY_ID = 12222222;
+    private QuestionGateway questionGateway;
+    private QuestionFactory questionFactory;
+    private ClientGateway clientGateway;
+    private AttorneyGateway attorneyGateway;
+    private TheQuestionOutputBoundary theQuestionOutputBoundary;
+    private QuestionInputBoundary questionInputBoundary;
+
+    public void setUpAskQuestionUseCase(){
```
Comment (yuzhouyi) [inline src/test/java/usecasetesting/AskQuestionUseCaseTest.java]: Based on my experience testing databases, I recommend to use the BeforeAll annotation for setUp methods to ensure they run before tests during runtime. The same could be said for all setUp methods.

File: src/test/java/usecasetesting/AskQuestionUseCaseTest.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        Attorney secondAttorney = new Attorney();
+        attorney.setUserId(SECOND_ATTORNEY_ID);
+        attorneyGateway.addUser(secondAttorney);
+    }
+
+    @Test
+    public void TestAskQuestionPassed(){
+        setUpAskQuestionUseCase();
+
+        QuestionRequestModel inputData = new QuestionRequestModel("fraud", "Test title", LocalDate.now(), CLIENT_ID, LocalDate.now());
```
Comment (yuzhouyi) [inline src/test/java/usecasetesting/AskQuestionUseCaseTest.java:81]: I believe we can further organize the tests by initiallizing the inputData after setting it as an instance variable. Then it would not need to be initialized after every test. The same could be said for all test methods.

File: src/test/java/usecasetesting/BrowseQuestionUseCaseTest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            assertEquals(2, response.getQuestionMap().size(), "The Question Map is not correct.");
+            List<QuestionDisplayFormatter> arrayList;
+            arrayList = new ArrayList<>(response.getQuestionMap().values());
+            assertEquals("test title", arrayList.get(0).getTitle());
+            return null;
+        }
+    };
+    private BrowseInputBoundary browseInputBoundary;
+
+    public void setUpBrowseUseCase(){
```
Comment (yuzhouyi) [inline src/test/java/usecasetesting/BrowseQuestionUseCaseTest.java:58]: I believe it is better to fully initiallize the entities instead of setting only the information you need. By fully initiallizing the entities, we can ensure extensibility for future code and testing, as well as better efficiency when running the code. Furthermore, by initiallizing all attributes instead of setting it one by one, we can better organize the code for easier reading and understanding. This can also be said for all other classes.

File: src/test/java/usecasetesting/CloseUseCaseTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        questionGateway.saveQuestion(question);
+    }
+    @Test
+    public void TestClientCloseableQuestion(){
+        setUpCloseUseCase();
+
+        CloseRequestModel inputData = new CloseRequestModel(QUESTION_ID, CLIENT_ID);
+
+        closeInputBoundary.closeQuestion(inputData);
+        Question question = questionGateway.getQuestion(QUESTION_ID);
```
Comment (yuzhouyi) [inline src/test/java/usecasetesting/CloseUseCaseTest.java:95]: I recommend including tests can verify the changes to the database, as that would help catch potential errors in the use case classes. This is highly important to the overall functionality of the code as a result of the nature of our program. Being able to catch cases where certain attributes of entities fail to be updated is extremely important.

File: src/test/java/usecasetesting/ReplyUseCaseTest.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        postInputBoundary.createPost(inputData2);// AttorneyQuestionList is not updated
+
+        User user = attorneyGateway.getUser(ATTORNEY_ID);
+        Question attorneyquestion = user.getQuestionsList().get(0);
+        assertEquals(attorneyquestion.getQuestionId(), QUESTION_ID);
+        assertEquals(attorneyquestion.getTakenByAttorney(), ATTORNEY_ID);
+        assertEquals(attorneyquestion.isTaken(), true);
+    }
+
+}
```
Comment (yuzhouyi) [inline src/test/java/usecasetesting/ReplyUseCaseTest.java:175]: I recommend adding a tearDown class with anotation AfterAll that deletes all entities that were added during setup and testing, as allowing them to stay in the database may result in problems in the future as well as taking unnecessary space.

File: src/test/java/usecasetesting/ClientRegisterUseCaseTest.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        };
+
+        clientRegisterInputBoundary = new ClientRegisterInteractor(clientGateway, registerOutputBoundary, clientFactory);
+
+    }
+    @Test
+    public void TestSuccessfulRegistration(){
+        setUpClientRegisterUseCase();
+
+        ClientRegisterRequestModel inputData = new ClientRegisterRequestModel("joseph", "test@gmail.com", PASSWORD, PASSWORD, "test state", "21000", "test ethinicity", 18, "male", "single", 1, 1000.0F);
```
Comment (yuzhouyi) [inline src/test/java/usecasetesting/ClientRegisterUseCaseTest.java:67]: I believe it is better practice to set these values as instance variables instead of slapping the raw value straight into the initialization.

File: src/test/java/usecasetesting/ReplyUseCaseTest.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        Question attorneyquestion = user.getQuestionsList().get(0);// AttorneyQuestionList is not updated
+        assertEquals(attorneyquestion.getQuestionId(), QUESTION_ID);
+        assertEquals(attorneyquestion.getTakenByAttorney(), ATTORNEY_ID);
+        assertEquals(attorneyquestion.isTaken(), true);
+
+        PostRequestModel inputData2 = new PostRequestModel(QUESTION_ID, ATTORNEY_ID, "Test text2");// Attorney
+        postInputBoundary.createPost(inputData2);
+        Question question = questionGateway.getQuestion(QUESTION_ID);
+        Post post2 = question.getPosts().get(1);
+        assertEquals(post2.getBelongsTo(), ATTORNEY_ID);
```
Comment (yuzhouyi) [inline src/test/java/usecasetesting/ReplyUseCaseTest.java:132]: Could be nice to include a message for the assertions to make sure we know where the failure occurrs if there is one.

**Review by yuzhouyi — CHANGES_REQUESTED** [review]
Overall, Joseph's test are comprehensive and I believe it covers all the important aspects of the use cases. However, I believe there are infact areas that require changing. I have left my recommendations as comments within the code, most being applicable to all classes. Additionally, based on my experience testing databases, I highly recommend including tests that checks for potential changes in the database, which can help catch errors in the use case class, especially the failure to update certain attributes of entities. As a result of the nature of our program, this is highly important, and thus I am requesting changes before approving this pull request.

**Review by YifanLiu2 — APPROVED** [review]

**Review by yuzhouyi — APPROVED** [review]
